xla/service/gpu/gemm_algorithm_picker_test.cc

```diff
@@ -52,31 +53,102 @@ class GemmAlgorithmPickerTest : public HloTestBase,
+      [](const se::CudaComputeCapability& cc) {
+        if(cc.IsAtLeastAmpere()) {
+          GTEST_SKIP() << "Skipping this test for Ampere+ as it is supported and "
                     "recommended with "
                     "the Nvidia Volta+ GPUs.";
+        }
+      },
+      [this](const se::RocmComputeCapability& cc) {
+        if(GetDebugOptionsForTest().xla_gpu_enable_cublaslt() &&
+          !cc.has_hipblaslt()) {
+          GTEST_SKIP() << "No gpublas-lt support on this architecture!";
+        }
+      }},
+      gpu_comp());
+  }
+};
+
+TEST_P(GemmAlgorithmPickerTest, SkipAlgorithmsWithAccuracyCheck) {
+
+  constexpr absl::string_view kHlo = R"(
+HloModule module
+
+ENTRY main {
+  %arg0 = f32[100,100]{1,0} parameter(0)
+  %arg1 = f32[100,100]{1,0} parameter(1)
+  ROOT %dot = f32[100,100]{1,0} dot(arg0, arg1), lhs_contracting_dims={1}, rhs_contracting_dims={0}
+})";
+
+  auto module_cfg = GetModuleConfigForTest();
+  auto debug_opts = module_cfg.debug_options();
+  size_t num_left1 = 0, num_left2 = 0;
+
+  {
+    // Run first with default settings (autotune level = 4), keep the number of
+    // algorithms left after autotuning
+    TF_ASSERT_OK_AND_ASSIGN(auto m,
+                          ParseAndReturnVerifiedModule(kHlo, module_cfg));
+    TF_ASSERT_OK_AND_ASSIGN(
+      bool changed,
+      RunHloPass(
+          GemmRewriter(gpu_comp(), /*toolkit_version=*/12040),
+          m.get()));
+
+    AutotuneConfig cfg{DeviceConfig{stream_exec(), nullptr}, debug_opts};
+    GemmAlgorithmPicker gpicker(cfg);
+    TF_ASSERT_OK_AND_ASSIGN(changed, RunHloPass(gpicker, m.get()));
+    ASSERT_TRUE(changed);
+    num_left1 = gpicker.num_algorithms_left();
+    if(num_left1 < 2) {
+      GTEST_SKIP() << "Too few algorithms left after the first step";
+    }
+  }
+
+  // Clear cache before the second run!
+  AutotunerUtil::ClearAutotuneResults();
+  {
+    // Run once again but now with autotune level 5 and embarassignly tight 
+    // rtol which shall disqualify most of the algorithms.
+
+    // Note that, we have "two sources of truth" for GemmAlgorithmPicker: i.e.,
+    // debug_options are used to initialize both 'HloModuleConfig' and also 
+    // 'AutotuneConfig'..
+    debug_opts.set_xla_gpu_autotune_gemm_rtol(1e-12);
+    debug_opts.set_xla_gpu_autotune_level(5);
+    module_cfg.set_debug_options(debug_opts);
+    TF_ASSERT_OK_AND_ASSIGN(auto m,
+                          ParseAndReturnVerifiedModule(kHlo, module_cfg));
+    TF_ASSERT_OK_AND_ASSIGN(
+      bool changed,
+      RunHloPass(
+          GemmRewriter(gpu_comp(), /*toolkit_version=*/12040),
+          m.get()));
+
+    AutotuneConfig cfg{DeviceConfig{stream_exec(), nullptr}, debug_opts};
+    GemmAlgorithmPicker gpicker(cfg);
+    TF_ASSERT_OK_AND_ASSIGN(changed, RunHloPass(gpicker, m.get()));
+    ASSERT_TRUE(changed);
+    num_left2 = gpicker.num_algorithms_left();
   }
+  // Assert that we have much fewer algorithms left after the second run
```

Comment:
nit: remove the word "much". We only assert that we have fewer algorithms. End the line with '.'